llvm/include/llvm/IR/ConstantRangeList.h

```diff
@@ -0,0 +1,86 @@
+//===- ConstantRangeList.h - A list of constant ranges ----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Represent a list of signed ConstantRange and do NOT support wrap around the
+// end of the numeric range. Ranges in the list are ordered and not overlapping.
+// Ranges should have the same bitwidth. Each range's lower should be less than
+// its upper.
```

Comment:
Something I noticed is that while ConstantRangeList maintains the "ordered and not overlapping" invariant through manipulations, it does not check that it initially holds. And indeed, LLParser will create such invalid ConstantRangeLists, which will later get rejected by the Verifier. Is this how it is *intended* to work out?

Reply:
it is convenient that the verifier can check for sorted `ConstantRangeList`s without asserting on construction, but perhaps it is better to move the checks to both LLParser and bitcode parser

a static method that returns a `std::optional<ConstantRangeList>` depending on if the input is sorted or not, and error checking in the LLParser/bitcode parser would be good. then we can assert that the input is sorted in the constructor.

(or just automatically sort ConstantRangeLists on construction, but IMO the checks can be useful for finding bugs)

Reply:
Added `static std::optional<ConstantRangeList> getConstantRangeList()` and checked the optional return in LLParse and bitcode parser.